[v6r22] Rucio File Catalog prototype

cserf opened the pull request:
BEGINRELEASENOTES

*Resources
NEW: RucioFileCatalog plugin based on Belle II implementation

ENDRELEASENOTES

fstagni commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+from __future__ import print_function
```
Add a docstring

petricm replied:
And in general all docstrings in this file also need variable explanations e.g.
```
:param x: is a good variable
```

fstagni commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+from stat import *
+import os
+import re
+import time
```
`time` and re` are unused.

fstagni commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+    """
+    successful, failed = {}, {}
+    lfnChunks = breakListIntoChunks(lfns, 1000)
+    listFiles = deepcopy(lfns.keys())
```
```suggestion
    listFiles = deepcopy(list(lfns))
```

fstagni commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+      proxyInfo = getProxyInfo(disableVOMS=True)
+      if proxyInfo['OK']:
+        os.environ['X509_USER_PROXY'] = proxyInfo['Value']['path']
+        gLogger.debug('X509_USER_PROXY not defined. Using %s' % proxyInfo['Value']['path'])
```
Don't use directly `gLogger`. Instead use
```
self.log = gLogger.getSubLogger('RucioFileCatalog')
```
and then use `self.log`

chaen replied:
Actually, we agreed to rather do `sLog = gLogger.getSubLogger(__name__)` as a global variable and use that

fstagni commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+      self.scopes = self.client.list_scopes()
+      self.account = self.client.account
+    except Exception as err:
+      gLogger.error('Cannot instantiate RucioFileCatalog interface, error : %s' % str(err))
```
```suggestion
      self.log.error('Cannot instantiate RucioFileCatalog interface', 'error : %s' % repr(err))
```

fstagni commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+    """ Check if the path exists
+    """
+    result = {'OK': True, 'Value': {'Successful': {}, 'Failed': {}}}
+    self.client = Client()
```
Do you really need to re-instantiate the `Client` object at every method call?

cserf replied:
Unfortunately yes. When one instantiates a `Client`, it creates a http session. This session will expire after some time, then it's not possible to do any new call.

chaen replied:
If you know the lifetime of the session, you can safely use the `DictCache` object

chrisburr replied:
If you can check if the connection is still alive you could use a property:
```python
  @property
  def client(self):
    if not self._client.is_alive:
      self._client = Client()
    return self._client
```

cserf replied:
Done

fstagni commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+        for lfn in list(lfnList):
+          successful[lfn] = True
+      except Exception as err:
+        # Try inserting one by one
```
add a `self.log.warn()` ?

fstagni reviewed (commented):
You need to add `rucio` to the requirements.txt file

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+from copy import deepcopy
+
+import DIRAC
+from DIRAC                                              import gConfig
```
'DIRAC.gConfig' imported but unused

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+from DIRAC                                              import gConfig
+from DIRAC                                              import S_OK, S_ERROR, gLogger
+from DIRAC.Resources.Catalog.Utilities                  import checkCatalogArguments
+from DIRAC.Core.Utilities.Time                          import fromEpoch
```
'DIRAC.Core.Utilities.Time.fromEpoch' imported but unused

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+from DIRAC.Resources.Catalog.Utilities                  import checkCatalogArguments
+from DIRAC.Core.Utilities.Time                          import fromEpoch
+from DIRAC.Core.Utilities.List                          import breakListIntoChunks
+from DIRAC.Core.Security.ProxyInfo                      import getProxyInfo, formatProxyInfoAsString
```
'DIRAC.Core.Security.ProxyInfo.formatProxyInfoAsString' imported but unused

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+from DIRAC.Core.Utilities.Time                          import fromEpoch
+from DIRAC.Core.Utilities.List                          import breakListIntoChunks
+from DIRAC.Core.Security.ProxyInfo                      import getProxyInfo, formatProxyInfoAsString
+from DIRAC.ConfigurationSystem.Client.Helpers.Registry  import getDNForUsername, getVOMSAttributeForGroup, \
```
'DIRAC.ConfigurationSystem.Client.Helpers.Registry.getDNForUsername' imported but unused
'DIRAC.ConfigurationSystem.Client.Helpers.Registry.getVOMSAttributeForGroup' imported but unused
'DIRAC.ConfigurationSystem.Client.Helpers.Registry.getVOForGroup' imported but unused
'DIRAC.ConfigurationSystem.Client.Helpers.Registry.getVOOption' imported but unused

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+
+from rucio.client import Client
+from rucio.common.exception import DataIdentifierNotFound, DuplicateContent, FileReplicaAlreadyExists, \
+                                   FileAlreadyExists, CannotAuthenticate, MissingClientParameter
```
'rucio.common.exception.DuplicateContent' imported but unused
'rucio.common.exception.FileAlreadyExists' imported but unused
'rucio.common.utils.chunks' imported but unused

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+        gLogger.debug('X509_USER_PROXY not defined. Using %s' % proxyInfo['Value']['path'])
+    try:
+      self.client = Client()
+    except (CannotAuthenticate, MissingClientParameter) as err:
```
err assigned but not used, maybe propagate the content of the exception?

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+                if self.convertUnicode:
+                  pfn = str(pfn)
+                result['Value']['Successful'][lfn]['Files'][name]['Replicas'][rse] = {'PFN': pfn, 'Status': 'U'}
+      except DataIdentifierNotFound as err:
```
err assigned but not used, maybe propagate the content of the exception?

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+from __future__ import division
+__RCSID__ = "$Id"
+
+from stat import *
```
I don't think `import *` is at anytime a good idea as we don't know what is imported, but anyway `stat.*` is not used anywhere

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+
+from stat import *
+import os
+import re
```
unused

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+      gLogger.error('Cannot instantiate RucioFileCatalog interface, error : %s' % str(err))
+
+
+  def __getDidsFromLfn(self, lfn):
```
docstring

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+  ADMIN_METHODS = FileCatalogClientBase.ADMIN_METHODS + ['getUserDirectory', 'createUserDirectory',
+                                                         'createUserMapping', 'removeUserDirectory']
+
+  def __init__(self, **options):
```
docstring

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+
+
+  @checkCatalogArguments
+  def getCompatibleMetadata(self, queryDict, path, credDict):
```
credDict and queryDict unused

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+    self.client = Client()
+    for lfnList in lfnChunks:
+      try:
+        DidList = [self.__getDidsFromLfn(lfn) for lfn in lfnList if lfn]
```
DidList -> didList wrong variable name style

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+
+
+  @checkCatalogArguments
+  def getReplicas(self, lfns, allStatus = False):
```
allStatus not used

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+
+
+  @checkCatalogArguments
+  def listDirectory(self, lfns, verbose = False):
```
verbose unused

petricm replied:
this function has a lot of if, it would be good to split it up or put some documentation in between at least

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+
+
+  @checkCatalogArguments
+  def getFileMetadata(self, lfns, ownership = False):
```
ownership unused

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+    result = {'OK': True, 'Value': {'Successful': {}, 'Failed': {}}}
+    self.client = Client()
+    for lfn in lfns:
+      size = 0
```
unused size

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+
+
+  @checkCatalogArguments
+  def addFile(self, lfns):
```
docstring

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+
+
+  @checkCatalogArguments
+  def removeReplica(self, lfns):
```
docstring

petricm commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,545 @@
+
+
+  @checkCatalogArguments
+  def getDirectorySize(self, lfns, longOutput=False, rawFiles=False):
```
rawFiles unussed

petricm replied:
longOutput unused

petricm commented:
This also requires addition of `rucio` to:
 - environment.yml
 - environment-py3.yml
 - requirements.txt
 - DIRACOS

chrisburr commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+                            continue
+
+                        # Add RSE attributes for the new RSE
+                        dictRSEAttributes = {'naming_convention': 'BelleII', 'ANY': True, 'fts': 'https://fts.usatlas.bnl.gov:8446'}
+                        dictRSEAttributes = {'naming_convention': 'BelleII', 'ANY': True, 'fts': 'https://kek2-fts01.cc.kek.jp:8446'}
```
I guess this should come from the CS?

cserf replied:
Yes. Any suggestion in which section ?

fstagni replied:
The list of FTS servers is `/Resources/FTSEndpoints`

chrisburr commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+                        protocols_to_create.append(prot)
+                        if prot not in existing_protocols and prot[0] in valid_protocols:
+                            # The protocol defined in Dirac does not exist in Rucio. Will be created
+                            self.log.info('Will create new protocol %s://%s:%s%s on %s' % (params['scheme'], params['hostname'], params['port'], params['prefix'], se))
```
You should run the linting from one of the newer branches, as a few lines are far too long

cserf replied:
Done

chrisburr commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+                            for protocol in protocols:
+                                if prot == (str(protocol['scheme']), str(protocol['hostname']), str(protocol['port']), str(protocol['prefix'])):
+                                    # Check if the protocol defined in Dirac has the same priority as the one defined in Rucio
+                                    for domain in ['lan', 'wan']:
+                                        for activity in ['read', 'write', 'delete']:
+                                            if params['domains'][domain][activity] != protocol['domains'][domain][activity]:
```
Can this just be `if params['domains'] != protocol['domains']:` or are there other keys in `domains` that need to be ignored?

chrisburr commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+            rses = [rse['rse'] for rse in client.list_rses()]
+            if newRSE:
+                self.log.info("Adding distances")
+                for src_rse in rses:
+                   for dest_rse in rses:
+                       if src_rse != dest_rse:
+                           try:
+                               client.add_distance(src_rse, dest_rse, {'ranking': 1, 'distance': 10})
+                           except Exception as err:
+                               self.log.error('Cannot add distance for %s:%s : %s' % (src_rse, dest_rse, str(err)))
```
```suggestion
                for src_rse, dest_rse in itertools.permutations(rses, r=2):
                   try:
                       client.add_distance(src_rse, dest_rse, {'ranking': 1, 'distance': 10})
                   except Exception as err:
                       self.log.error('Cannot add distance for %s:%s : %s' % (src_rse, dest_rse, str(err)))
```

cserf replied:
:+1:

chrisburr commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+            if result['OK']:
+                for dataLevel in result['Value']:
+                    result = Operations().getOptionsDict('Shares/%s' % dataLevel)
+                    if result['OK']:
```
Another example where `continue` can reduce indentation and make it easier to follow

cserf replied:
I don't understand the comment

chrisburr replied:
Something like this to avoid indenting any more deeply than needed:
```python
if not result['OK']:
  self.log.error('Cannot get SEs : %s' % result['Value'])
  continue
attribute = '%sShare' % dataLevel
for rse in rses:
  value = result['Value'].get(rse, 0)
  self.log.info('Setting %s for %s : %s' % (attribute, rse, value))
  try:
    client.add_rse_attribute(rse, attribute, value)
  except Exception as err:
    self.log.error('Cannot create %s for %s' % (attribute, rse))
```

chrisburr replied:
Also I guess the logging should used the fixed/variable split like:
```diff
- self.log.error('Cannot create %s for %s' % (attribute, rse))
+ self.log.error('Cannot create RSE attribute: ', '%s for %s' % (attribute, rse))
```

cserf replied:
Ah, right thanks for the clarification

cserf replied:
Actually, we don't want to continue here because other things are done after and we don't want to skip them

chrisburr replied:
Nothing else is done in the `for dataLevel in result['Value']:` loop? Or am I missing something?

cserf replied:
Yes, you're right. Done

chrisburr commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+                self.log.info('Will set primarySEs flag to %s' % str(primarySEs))
+                for rse in rses:
+                    if rse in allSEs:
+                        storage = StorageElement(rse)
+                        OccupancyLFN = storage.options.get('OccupancyLFN')
+                        try:
+                            client.add_rse_attribute(rse, 'OccupancyLFN', OccupancyLFN)
```
```suggestion
                        occupancyLFN = StorageElement(rse).options.get('OccupancyLFN')
                        try:
                            client.add_rse_attribute(rse, 'OccupancyLFN', occupancyLFN)
```

cserf replied:
Done

chrisburr commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+            listAccounts = [str(acc['account']) for acc in client.list_accounts()]
+            listScopes = [str(scope) for scope in client.list_scopes()]
+            accounts = getAllUsers()
+            DNMapping = {}
```
This should start with a lower case character

cserf replied:
Fixed

chrisburr commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+            self.log.info("Synchronizing accounts")
+            listAccounts = [str(acc['account']) for acc in client.list_accounts()]
+            listScopes = [str(scope) for scope in client.list_scopes()]
+            accounts = getAllUsers()
```
I think there is no need for this to be a variable, it can simply be `for account in getAllUsers():`

cserf replied:
Done

chrisburr commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+            # Collect the group accounts from Dirac Configuration and create service accounts in Rucio
+            groups = getAllGroups()
+            for group in groups:
+                dns = getDNsInGroup(group)
```
This variable looks like it can also be dropped in favour of using `for dn in getDNsInGroup(group)`

cserf replied:
Done

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+sLog = gLogger.getSubLogger(__name__)
+
+
+def get_scope(lfn, scopes=[]):
+  """Helper function that extract the scope from the LFN
+  """
+  scope, _ = extract_scope(did=lfn, scopes=scopes)
```
Even if `scopes` is never modified it's safer to do this:
```suggestion
def get_scope(lfn, scopes=None):
  """Helper function that extract the scope from the LFN
  """
  if scopes is None:
    scopes = []
  scope, _ = extract_scope(did=lfn, scopes=scopes)
```

cserf replied:
Done

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+    """
+    try:
+      self._client.ping()
+      return self._client
+    except:
+      self._client = Client(account=self.account)
+      self.scopes = self._client.list_scopes()
+      return self._client
```
```suggestion
    except Exception:
      self._client = Client(account=self.account)
      self.scopes = self._client.list_scopes()
    return self._client
```

cserf replied:
Done

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+    """ Helper function to convert LFNs into DIDs
+    """
+    if lfn.find(':') > -1:
+      scope, name = lfn.split(':')
+      return {'scope': scope, 'name': name}
+    else:
+      scope = get_scope(lfn, scopes=self.scopes)
+    return {'scope': scope, 'name': lfn}
```
```suggestion
      scope, name = lfn.split(':')
    else:
      scope = get_scope(lfn, scopes=self.scopes)
      name = lfn
    return {'scope': scope, 'name': name}
```

cserf replied:
Done

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+    """ Returns replicas for an LFN or list of LFNs
+    """
+    result = {'OK': True, 'Value': {'Successful': {}, 'Failed': {}}}
+    lfnChunks = breakListIntoChunks( lfns, 1000 )
+
+    for lfnList in lfnChunks:
```
No need for the `lfnChunks` variable (especially as I think this produced a generator so reusing `lfnChunks` will give wrong results).

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+  def __init__(self, **options):
+    """ Constructor
+    """
+    self.convertUnicode = True
```
Can you explain what this is trying to achieve? In particular is `str(...)` the right thing to do for Python 3 as well? Or should it be `.encode()`?

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+          for rep in self.client.list_replicas([did, ]):
+            if rep:
+              name = rep['name']
+              if self.convertUnicode:
```
Subject to the outcome of my other Unicode comment, this feels like it should be a function so you don't have to full the code with `if`. Or better yet always use `str` as I think `convertUnicode` is always True?

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+    successful, failed = {}, {}
+    lfnChunks = breakListIntoChunks(lfns, 1000)
+    listFiles = deepcopy(list(lfns))
+    for chunk in lfnChunks:
```
No need for a `lfnChunks` variable

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+                               'NumberOfLinks': 1,
+                               'Size': meta['bytes'],
+                               'Status': '-'}
+            try:
+              listFiles.remove(lfn)
+            except ValueError:
+              pass
```
These doesn't need to be duplicated as the same thing is done for both branches of the `if`

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+    """ Determine whether the path is a directory
+    """
+    result = {'OK': True, 'Value': {'Successful': {}, 'Failed': {}}}
+    dids = []
+    for lfn in lfns:
+      dids.append(self.__getDidsFromLfn(lfn))
```
```suggestion
    dids = [self.__getDidsFromLfn(lfn) for lfn in lfns]
```

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+    """
+
+    result = {'OK': True, 'Value': {'Successful': {}, 'Failed': {}}}
+    dids = []
+    for lfn in lfns:
+      dids.append(self.__getDidsFromLfn(lfn))
```
```suggestion
    dids = [self.__getDidsFromLfn(lfn) for lfn in lfns]
```

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+          result['Value']['Successful'][lfn] = True
+        else:
+          result['Value']['Successful'][lfn] = False
+      for lfn in lfns:
+        if lfn not in result['Value']['Successful'] and lfn not in result['Value']['Failed']:
+          result['Value']['Failed'][lfn] = 'No such file or directory'
```
I think there is no need for this to be inside the `try...except`

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+    try:
+      for meta in self.client.get_metadata_bulk(dids):
+        lfn = str(meta['name'])
+        if meta['did_type'] in ['FILE']:
+          result['Value']['Successful'][lfn] = True
+        else:
+          result['Value']['Successful'][lfn] = False
```
```suggestion
        result['Value']['Successful'][lfn] = (meta['did_type'] == 'FILE')
```

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+        rep = {'scope': did['scope'], 'name': did['name'], 'bytes': size, 'adler32': checksum}
+        if pfn:
+          rep['pfn'] = pfn
+        self.client.add_replicas(rse=se, files=[rep, ])
```
Is there a reason for the `, `?

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+    try:
+      for meta in self.client.get_metadata_bulk(dids):
+        lfn = str(meta['name'])
+        if meta['did_type'] in ['DATASET', 'CONTAINER']:
+          result['Value']['Successful'][lfn] = True
+        else:
+          result['Value']['Successful'][lfn] = False
```
```suggestion
        result['Value']['Successful'][lfn] = (meta['did_type'] in ['DATASET', 'CONTAINER'])
```

cserf replied:
Done

cserf replied:
Done

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+    """ Determine whether the path is a file
+    """
+
+    result = {'OK': True, 'Value': {'Successful': {}, 'Failed': {}}}
```
And then return S_OK?
```suggestion
    result = {'Successful': {}, 'Failed': {}}
```

cserf replied:
Done

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+      return S_ERROR(str(err))
+    return result
+
+
```
Too much whitespace, if you run the linter or open a draft PR against a newer branch you'll see some other errors as well
```suggestion
```

chrisburr commented on Resources/Catalog/RucioFileCatalogClient.py:
```diff
@@ -0,0 +1,590 @@
+            resDict['Successful'][lfn]['Files'] += 1
+            resDict['Successful'][lfn]['TotalSize'] += child['bytes']
+          for rep in self.client.list_replicas([{'scope': did['scope'], 'name': did['name']}]):
+            fileSize = rep['bytes']
+            for rse in rep['rses']:
+              if rse not in resDict['Successful'][lfn]['SiteUsage']:
+                resDict['Successful'][lfn]['SiteUsage'][rse] = {'Files': 0, 'Size': 0}
+              resDict['Successful'][lfn]['SiteUsage'][rse]['Files'] += 1
+              resDict['Successful'][lfn]['SiteUsage'][rse]['Size'] += fileSize
```
Can this be replaced by:
```suggestion
            resDict['Successful'][lfn]['SiteUsage'][rse] = {
                'Files': len(rep['rses']),
                'Size': len(rep['rses']) * rep['bytes'],
            }
```

fstagni commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+"""
+
+# # imports
+from __future__ import print_function
```
Add also
```
from __future__ import division
from __future__ import absolute_import
```

cserf replied:
Done

fstagni commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+    seProtocols = {}
+    dms = DMSHelpers()
+    for seName in dms.getStorageElements():
+        seProtocols[seName] = []
```
Leave that fight for another day ;)

fstagni commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+    return S_OK(seProtocols)
+
+
+class RucioSynchronizer(AgentModule):
```
We always names the agents with "blaBlaAgent", and with the notable exception of the SiteDirector, this is a convention everywhere. So, this class, and the file name, should be "RucioSynchronizerAgent".

fstagni commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+        valid_protocols = ['srm', 'gsiftp', 'davs', 'https', 'root']
+        default_email = None
+        try:
+            client = Client(account='root', auth_type='userpass')
```
Where are these coming from?

cserf replied:
Rucio config file.

fstagni commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+        :param self: self reference
+        """
+
+        valid_protocols = ['srm', 'gsiftp', 'davs', 'https', 'root']
```
Can we have a non-hardcoded list?

fstagni commented on ConfigurationSystem/Agent/RucioSynchronizer.py:
```diff
@@ -0,0 +1,359 @@
+                    for prot in protocols:
+                        existing_protocols.append((str(prot['scheme']), str(prot['hostname']), str(prot['port']), str(prot['prefix'])))
+                    protocols_to_create = []
+                    for cnt in range(len(result['Value'][se])):
```
use `enumerate`?

fstagni commented on ResourceStatusSystem/Agent/RucioRSSAgent.py:
```diff
@@ -0,0 +1,69 @@
+    try: 
+      for rse in client.list_rses():
+        thisSe = rse['rse']
+        gLogger.info("Running on %s" % thisSe)
```
```suggestion
        gLogger.info("Running on", thisSe)
```

fstagni commented on ResourceStatusSystem/Agent/RucioRSSAgent.py:
```diff
@@ -0,0 +1,69 @@
+          seAccessValue = resStatus['Value'][thisSe]
+          availabilityRead = True if seAccessValue['ReadAccess'] == 'Active' else False
+          availabilityWrite = True if seAccessValue['WriteAccess'] == 'Active' else False
+          availabilityDelete = True if seAccessValue['RemoveAccess'] == 'Active' else False
```
RSS has actually 4 states, you states, you can read about them in https://dirac.readthedocs.io/en/latest/AdministratorGuide/Systems/ResourceStatus/introduction.html#state

So I think you can do 

```suggestion
          availabilityRead = True if seAccessValue['ReadAccess'] in ('Active', 'Degraded') else False
          availabilityWrite = True if seAccessValue['WriteAccess'] in ('Active', 'Degraded') else False
          availabilityDelete = True if seAccessValue['RemoveAccess'] in ('Active', 'Degraded') else False
```

cserf replied:
Done

fstagni commented on ResourceStatusSystem/Agent/RucioRSSAgent.py:
```diff
@@ -0,0 +1,69 @@
+
+    :param self: self reference
+    """
+    gLogger.info("Starting RucioRSSAgent")
```
Don't use `gLogger` directly, use `self.log`

cserf replied:
Done

fstagni commented on ResourceStatusSystem/Agent/RucioRSSAgent.py:
```diff
@@ -0,0 +1,69 @@
+
+    Agent that synchronizes Rucio and Dirac
+"""
+
```
Add the `from __future__ import ...`

martynia commented:
https://github.com/DIRACGrid/DIRAC/blob/85478950347ee216b99a00643285a2ceccfed305/Resources/Catalog/RucioFileCatalogClient.py#L58-L86

You are passing the `**options` argument but it is not used. The `FileCatalog` dispatcher passes options defined in the CS to the client. These could be used to configure the rucio client and should in my opinion be given a preference over a config file which the `client()` would use internally. This way by passing the parameters from `options` to `client()` we could eliminate a need of a config file entirely, at list for a minimal Dirac-rucio configuration. These options could include `rucio_host` and `auth_host` for example. The account and VO could come from a proxy.

fstagni commented:
One more thing: once the reviewer's comments have been addressed, I suggest to rebase this development on rel-v7r0. At a minimum you would get better automated checks, as on v6r22 branch there's not even GitHub Actions.

fstagni commented:
Hello, any updates?

cserf commented:
I addressed most of the comments. Thanks a lot for all the useful suggestions. I didn't implement yet passing all the options to the client constructor, because now it wouldn't work anyway without a config file. It can be improved later.
